File: Baekjoon/3000~3999/[3066]브리징 시그널/solution.py
import sys

T = int(sys.stdin.readline())
lis = []


def lower_bound(value):
    low = 0
    high = len(lis)

    while low < high:
        mid = (low + high) // 2

        if value < lis[mid]:
            high = mid
        else:
            low = mid + 1

    return high


while T > 0:
    N = int(sys.stdin.readline())

    for i in range(N):
        num = int(sys.stdin.readline())

        idx = lower_bound(num)

        if idx == len(lis):
            lis.append(num)
        else:
            lis[idx] = num

    sys.stdout.write(str(len(lis)) + "\n")

    lis.clear()
    T -= 1

File: Baekjoon/3000~3999/[3066]브리징 시그널/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

#define endl "\n"

vector<int> lis;
int T, N, num;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  cin >> T;

  while (T--) {
    cin >> N;

    for (int i = 0; i < N; i++) {
      cin >> num;

      int idx = lower_bound(lis.begin(), lis.end(), num) - lis.begin();

      if (idx == lis.size())
        lis.push_back(num);
      else
        lis[idx] = num;
    }

    cout << lis.size() << endl;

    lis.clear();
  }

  return 0;
}